docs(MADR): design SPIFFE compliance (#13804)
kumahq/kuma: docs/madr/decisions/073-spiffe-compliance.md

```
spiffe://<trustDomain>/ns/<namespace>/sa/<serviceAccount>
```

Trust domain: `{ .Mesh }.{ .Zone }.{ .ClusterID }.kuma.io`

Once a TrustDomain is defined on the Global control plane, we require it to include the zone name in the template. We are going to implement a validator for this case.

Question:
* Is there an issue once we rename zone ? Isn't it an edge case that we might don't need to support?
  It's not supported and can cause many issues, we shouldn't think about it.

### MeshIdentity

This resource allows the creation of Certificate Authorities (CAs) that are supported within a specific Mesh. It does not actively issue certificates — it simply defines the authority. The user must later configure which CA is used as the default, and which ones should continue to be supported during a migration period.

* The resource can be created on the Global control plane and synced to Zones.
* It can also be created on a Zone, in which case it will be synced to Global, but not propagated to other Zones.

KDS flags:
```golang
model.GlobalToZonesFlag | model.ZoneToGlobalFlag

> ```

Reviewer: Also want to say: https://github.com/kumahq/kuma/blob/e279bba108307837bbe966f032c5e95039759372/pkg/core/resources/model/resource.go#L216

It almost makes sense to show the entire type descriptor here (at least the non default fields).

**Model**
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels:
        app: test
  spiffeID: # optional
    trustDomain: "prod.example.fr" # { .Mesh }.{ .Zone }.{ .ClusterID}.kuma.io
    path: "/ns/{{ .Namespace }}/sa/{{ .ServiceAccount }}" # let's use the same as HostnameGenerator
  provider:
    type: builtin | provided | spire
    builtin:
      dataplaneCertificate:
        duration: 24h
    spire:
      agent:
        addresss: "/run/spire/sockets/agent.sock"
        timeout: 1s
    provided: # to extend in KM
      certificate:
        # one of
        secret:
        inline:
        inlineString:
      privateKey:
        # one of
        secret:

>         inline:

Author: Yes it makes sense. I based this on https://github.com/kumahq/kuma/blob/master/api/common/v1alpha1/datasource.go
```golang
// DataSource defines the source of bytes to use.
type DataSource struct {
	// Data source is a secret with given Secret key.
	Secret *string `json:"secret,omitempty"`
	// Data source is inline bytes.
	Inline *[]byte `json:"inline,omitempty"`
	// Data source is inline string`
	InlineString *string `json:"inlineString,omitempty"`
}
```

We might need a new type

Users can provide certificates in the following ways:

* Inline,
* File path,
* Secret reference.

This gives flexibility for organizations that already have an established PKI or need to integrate with existing certificate authorities.

```yaml
    provided: # to extend in KM
      certificate:
        # one of
        secret:
        inline:
        inlineString:
      privateKey:
        # one of
        secret:
        inline:
        inlineString:
      dataplaneCertificate:
        duration: 24h
```

**Spire**

The spire provider is a new type that configures Envoy to request its identity certificate directly from SPIRE.
In this first iteration:

* Workloads must be registered in SPIRE manually or by SPIRE controller.
* The control plane does not handle workload registration in SPIRE yet.

This approach allows Kuma to leverage SPIFFE-compliant identities issued by SPIRE while keeping the initial integration simple.

```yaml
    spire:
      agent:
        addresss: "/run/spire/sockets/agent.sock"
        timeout: 1s
``` 

Another thing is that we need to mount a volume with Spire Agent unix socket.

> 1. Add a new section to the injector configuration:

Author: Yes, it maybe makes more sense to have it in configuration of the control-plane to avoid potential issues

```yaml
runtime:
  kubernetes:
    injector:
      spire:
        path: /run/spire/sockets
        mountPath: /run/spire/sockets
```

If the `spire.mountPath` and `spire.path` are set the injector automatically:

* creates volume
```yaml
volumes:
  - name: spire-agent-socket
    hostPath:
      path: /run/spire/sockets
      type: Directory
```

* mount volume
```yaml
volumeMounts:
  - name: spire-agent-socket
    mountPath: /run/spire/sockets
    readOnly: true
```
* and inject label

```yaml
kuma.io/spire-integration: enabled
```

Purpose of the Label:
* This label signals that the pod has SPIRE integration enabled.

> * It prevents Kuma from delivering Spire-related configuration (e.g., SDS config, identity provider references) to pods that don't have the socket mounted.

Author: The only issue is that once you enable spire and you don't have a mounted volume it will send config which will be rejected

> * It prevents Kuma from delivering Spire-related configuration (e.g., SDS config, identity provider references) to pods that don't have the socket mounted.

Author: In a way, it makes sense - you want SPIRE, and the configuration is being delivered. However, if the required volume is missing, the pod isn't actually ready. This mechanism provides a safer way to enable SPIRE without breaking existing running workloads. It might not be required during first iteration but I think it's nice to have

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: old-identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels: {} # select all
....
---
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: new-identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels:
        app: demo-app
```

The `MeshIdentity` resource’s selector field allows selection based on dataplane labels. Users can define it to apply to either all dataplanes or a specific subset.
If there are multiple (X) identities selecting the same dataplane, the control plane will resolve the conflict by choosing one based on the following order of precedence:

1. Selector specificity – the identity with most specific selector is preferred.
2. Lexicographical order – if selector sizes are equal, the identity with the lexicographically smallest name is chosen.

> 3. Age – if names are the same, the oldest identity (based on creation timestamp) is selected.

Reviewer: Name can't be the same right?

#### How would look the process of switching backends?

1. User has a Mesh with MeshIdentity `ca-1`
```yaml
...
name: demo-client-65f94cb577-jxrrq-44wfv42db9z44xxf
type: DataplaneInsight
mTLS:
  certificateExpirationTime: "2025-06-03T17:35:33Z"
  certificateRegenerations: 1
  issuedBackend: ca-1 # name of MeshIdentity
  lastCertificateRegeneration: "2025-06-02T17:35:33.624189876Z"
  supportedBackends: # list of MeshTrust name
  - ca-1
...
```
2. User creates MeshTrust with CA of a new MeshIdentity `ca-2`
3. Trust is propagated. We provide identity and keep the information(supported CA (Trust), and Identity) in DataplaneInsight
```yaml
...
name: demo-client-65f94cb577-jxrrq-44wfv42db9z44xxf
type: DataplaneInsight
mTLS:
  certificateExpirationTime: "2025-06-03T17:35:33Z"
  certificateRegenerations: 1
  issuedBackend: ca-1 # name of MeshIdentity
  lastCertificateRegeneration: "2025-06-02T17:35:33.624189876Z"

>   supportedBackends: # list of MeshTrust name

Reviewer: KRI no?

  - ca-1
  - ca-2
...
```

Based on this information we will update MeshService

```yaml
...
name: second-test-server-b7dwz26bb9b4z676
spec:
  identities:
  - type: SpiffeID

>     value: spiffe://example.com/zone/my-new-zone/ns/real-resource-ns/sa/default

Reviewer: needs updating at least to have valid example

  issuedBackend: ca-2 # name of MeshIdentity
  lastCertificateRegeneration: "2025-06-02T17:35:33.624189876Z"
  supportedBackends: # list of MeshTrust name
  - ca-1
  - ca-2
...
```
Update SpiffeID based on identities
```yaml
...
name: second-test-server-b7dwz26bb9b4z676
spec:
  identities:
  - type: SpiffeID
    value: spiffe://example.com/ns/real-resource-ns/sa/default
  - type: SpiffeID
    value: spiffe://example.fr/ns/real-resource-ns/sa/default
  ports:
  - appProtocol: http
    name: main
    port: 80
    targetPort: main
  selector:
    dataplaneTags:
      app: second-test-server
      k8s.kuma.io/namespace: real-resource-ns
  state: Available
status:
  ...
  tls:
    status: Ready
  vips:
  - ip: 10.43.245.159
type: MeshService
```

6. User can remove MeshIdentity `ca-1`
7. MeshTrust should be removed by the user

![Rotation flow](assets/073/rotation-flow.png)

#### Secret naming

We want to use a [KRI](070-resource-identifier.md) for resource naming. Since `MeshIdentity` creates a separate resource (Secret), it should be named based on the KRI to simplify correlation between the `Secret` and the `MeshIdentity` configuration.

Proposed name:

```yaml

> kri_mi_mesh-1_us-east-2_kuma-system_identity-1

Reviewer: not going to work on k8s as this is a secret so it needs to be the right DNS format

```

Also, we need to name the Spire agent cluster, in this case is it internal or maybe based on MeshIdentity:
Proposed name:
```yaml

> system_spire-agent

Reviewer: ```suggestion
system_sds-spire-agent
```

```

or 
```yaml

> kri_mi_mesh-1_us-east-2_kuma-system_spire-identity

Reviewer: I think @slonka left a path just for this: `system_kri_mi_mesh-1_us-east-2_kuma-system_spire-identity`

Reviewer: Also I think mid might be a better shortname no?

Reviewer: I think this whole part goes too fast. It covers 2 options 1 where we're talking about envoy secret resource name and the other is an envoy cluster for sds

> https://spiffe.io/docs/latest/deploying/spire_agent/#sds-configuration
> https://spiffe.io/docs/latest/microservices/envoy/#tls-certificates
#### MeshTrust

Based on the existing `MeshIdentity`, we should create a `MeshTrust` resource, which will be used to properly generate trust domains for other zones.
`MeshTrust` should be created in each zone and then synchronized with the global control plane.
1. Can be created on the Zone and synced to Global
2. Can be created on the Global and synced to all Zones
3. Can be created on the Zone and not synced to global by providing label `kuma.io/kds-sync: disabled`
4. Can be created on the Zone and synced to all zones (Future release) (`SyncedAcrossZonesFlag`)

We can describe it as KDS flag:

```golang
model.GlobalToZonesFlag | model.ZoneToGlobalFlag
```

There should also be an option to disable automatic generation in cases where the user does not want cross trust-domain traffic to be allowed by default.

**Model**
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrust
metadata:
  name: trust-hash
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:

>   ca: inline | secret | file

Reviewer: That's a caBundle right?

Reviewer: Btw the CA rotation without trustDomain change is it just a matter of having the old and new bundle and then removing the old once it's expired or if a user wants?

Reviewer: would it be useful to have an optional origin?

  trustDomain: "prod.zone-1.id.kuma.io"
```

The resource should also be manually creatable by the user. This enables users to define additional trust domains and allows the local sidecar to accept incoming mTLS traffic from services outside the mesh.
This functionality is essential for:

* Enabling ExternalClients,
* Accepting mTLS connections from external services, and
* Federating multiple separate clusters through shared trust.

**MeshTrust based on MeshIdentity**

> When a user creates a `MeshIdentity`, we will automatically create a corresponding `MeshTrust` resource using the CA and trust domain from that identity. We can implement a dedicated generator that creates the MeshTrust based on the `MeshIdentity`.

Reviewer: I suggested we'd make this optional in the MeshIdentity resource.

> To avoid issues where a user removes a `MeshIdentity` but an old trust remains in use, we will not remove a `MeshTrust` and we let user to clean it up. In case it's problematic for the user we can implement it later.

Reviewer: What happens with updates to MeshIdentity or deletion/recreation? Is the MeshTrust replaced? Updated with both?

        secret:
        path:
        envVar:
      dataplaneCertificate:
        duration: 24h
```

Currently, the `ServiceAccount` is not exposed in the `Dataplane` resource, which is problematic since we need to access this information. A potential solution is to add a dedicated field to the `Dataplane` object to store the associated `ServiceAccount`.

#### Provider types

There can be multiple difference identity providers but in the first iteration we would like to focus on:

* provided
* spire

**Provided**

The provided type allows users to supply their own CA certificates. These can be configured per zone or as a global definition.
Users can provide certificates in the following ways:

* Environment variable,
* File path,
* Secret reference.

This gives flexibility for organizations that already have an established PKI or need to integrate with existing certificate authorities.

```yaml
    provided: # to extend in KM
      insecureAutogenerate: false
      certificate:
        # one of
        secret:
        path:
        envVar:
      privateKey:
        # one of
        secret:
        path:
        envVar:
      dataplaneCertificate:
        duration: 24h
```

> Additionally, we could introduce a configuration option called `insecureAutogenerate`, which automatically generates a default CA and persists it in the store. This would functionally be equivalent to the current builtin CA.

Author: Yes, we should configure it to secure way but I feel we don't need to allow customization since if a user wants a more secure it can provide it's own CA. I will add more details what kind of CA is a default

We could introduce overlay to the control-plane injector config.

Example: 

```yaml
runtime:
  kubernetes:
    injector:
        overlay:
            - path: spec.template.spec.volumes.spire-agent-socket
              op: add
              value:
                name: spire-agent-socket
                hostPath:
                  path: /run/spire/sockets
                  type: Directory
            - path: spec.template.spec.containers.kuma-sidecar.volumeMounts.spire-agent-socket
              op: add
              value:
                name: spire-agent-socket
                mountPath: /run/spire/sockets/
                readOnly: true          
```

#### Define active identity provider

Previously, defining the active identity for a group of services was done through a single, perhaps less flexible, mechanism. Now, the `MeshIdentity` resource, leveraging its `selector`, offers a more granular and controlled approach to activate specific configurations.

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: old-identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:

>       matchLabels: {} # select all

Reviewer 2: It would be nice to also cover if any of the following is allowed and what means if allowed:
```yaml
# no selector
spec:
  provider: ...
```
```yaml
spec:
  selector: {}
  provider: ...
```
```yaml
spec:
  selector:
     dataplane: {}
  provider: ...
```

metadata:
  name: new-identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels:
        app: demo-app
```

The `MeshIdentity` resource’s selector field allows selection based on dataplane labels. Users can define it to apply to either all dataplanes or a specific subset.
If there are multiple (X) identities selecting the same dataplane, the control plane will resolve the conflict by choosing one based on the following order of precedence:

1. Selector specificity – the identity with most specific selector is preferred.
2. Lexicographical order – if selector sizes are equal, the identity with the lexicographically smallest name is chosen.

#### How would look the process of switching backends?

1. User has a Mesh with MeshIdentity `identity-1`
```yaml
...
name: demo-client-65f94cb577-jxrrq-44wfv42db9z44xxf
type: DataplaneInsight
mTLS:
  certificateExpirationTime: "2025-06-03T17:35:33Z"
  certificateRegenerations: 1
  issuedBackend: system_kri_mi_mesh-1_us-east-2_kuma-system_identity-1
  lastCertificateRegeneration: "2025-06-02T17:35:33.624189876Z"
  supportedBackends: # list of MeshTrust name

>   - system_kri_mtrust_mesh-1_us-east-2_kuma-system_identity-1

Reviewer 2: why `system_`?

Author: https://github.com/kumahq/kuma/pull/13783/files#diff-f47b6cd23a182f9d0bd5449f425b8963b2267c38852faca3541d913f8334238dR198

...
```
2. User creates MeshTrust with CA of a new MeshIdentity `identity-2`
3. Trust is propagated. We provide identity and keep the information(supported CA (Trust), and Identity) in DataplaneInsight
```yaml
...
name: demo-client-65f94cb577-jxrrq-44wfv42db9z44xxf
type: DataplaneInsight
mTLS:
  certificateExpirationTime: "2025-06-03T17:35:33Z"
  certificateRegenerations: 1
  issuedBackend: system_kri_mi_mesh-1_us-east-2_kuma-system_identity-1 # name of MeshIdentity
  lastCertificateRegeneration: "2025-06-02T17:35:33.624189876Z"
  supportedBackends: # list of MeshTrust name
  - system_kri_mtrust_mesh-1_us-east-2_kuma-system_identity-1
  - system_kri_mtrust_mesh-1_us-east-2_kuma-system_identity-2
...
```

Based on this information we will update MeshService

```yaml
...

> name: second-test-server-b7dwz26bb9b4z676

Reviewer 2: In this example, does MeshService `second-test-server-b7dwz26bb9b4z676` select Dataplane  `demo-client-65f94cb577-jxrrq-44wfv42db9z44xxf`? or it's some other service that demo-client communicates with?

Author: Updated example with correct names

spec:
  identities:
  - type: SpiffeID

>     value: spiffe://old.zone-1.0c5afd2e-de93-437b-8ba0-3decdff805b6.kuma.io/ns/real-resource-ns/sa/default

Author: based on this yes https://github.com/kumahq/kuma/blob/master/pkg/clusterid/creator.go#L29-L45

  ports:
  - appProtocol: http
    name: main
    port: 80
    targetPort: main
  selector:
    dataplaneTags:
      app: second-test-server
      k8s.kuma.io/namespace: real-resource-ns
  state: Available
status:
  ...
  tls:

>     status: Ready <- only if all DPPs have identity provided

Reviewer 2: all DPPs have "identity provided" or "trust provided"?

Author: I this case we check only Identity Provided

>     status: Ready <- only if all DPPs have identity provided

Reviewer 2: Also I thought we don't want real traffic to be affected by the info we gathered through DataplaneInsights? But currently we use `status.tls.status` when configuring outbounds for MeshService

Author: We are using it now so I don't want to change this logic

  vips:
  - ip: 10.43.245.159
type: MeshService
```

4. User create a new MeshIdentity `identity-2`
5. We start deliver new identity provided by `identity-2`
We reuse existing fields
```yaml
...
name: demo-client-65f94cb577-jxrrq-44wfv42db9z44xxf
type: DataplaneInsight
mTLS:
  certificateExpirationTime: "2025-06-03T17:35:33Z"
  certificateRegenerations: 1
  issuedBackend: system_kri_mtrust_mesh-1_us-east-2_kuma-system_identity-2 # name of MeshIdentity
  lastCertificateRegeneration: "2025-06-02T17:35:33.624189876Z"
  supportedBackends: # list of MeshTrust name
  - system_kri_mtrust_mesh-1_us-east-2_kuma-system_identity-1
  - system_kri_mtrust_mesh-1_us-east-2_kuma-system_identity-2
...
```
Update SpiffeID based on identities
```yaml
...
name: second-test-server-b7dwz26bb9b4z676
spec:
  identities:
  - type: SpiffeID
    value: spiffe://old.zone-1.0c5afd2e-de93-437b-8ba0-3decdff805b6.kuma.io/ns/real-resource-ns/sa/default
  - type: SpiffeID

>     value: spiffe://new.zone-1.0c5afd2e-de93-437b-8ba0-3decdff805b6.kuma.io/ns/real-resource-ns/sa/default

Reviewer 2: why `old` and `new`? it's supposed to be a mesh name, no?

	WsPath:                       "meshtrusts",
	KumactlArg:                   "meshtrust",
	KumactlListArg:               "meshtrusts",
	AllowToInspect:               false,
	IsPolicy:                     false,
	IsExperimental:               false,
	SingularDisplayName:          "Mesh Trust",
	PluralDisplayName:            "Mesh Trusts",
	IsPluginOriginated:           true,
	IsTargetRefBased:             false,
	HasToTargetRef:               false,
	HasFromTargetRef:             false,
	HasRulesTargetRef:            false,
	HasStatus:                    false,
	AllowedOnSystemNamespaceOnly: true,
	IsReferenceableInTo:          false,
	ShortName:                    "mtrust",
	IsFromAsRules:                false,
}
```

There should also be an option to disable automatic generation in cases where the user does not want cross trust-domain traffic to be allowed by default.

**Model**
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrust
metadata:
  name: trust-hash
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  ca: inline | secret | file
  trustDomain: "prod.zone-1.id.kuma.io"
  originRef:
    kind: MeshIdentity
    name: default

>     namespace: kuma-system

Author: We don't sync across zones MeshTrust, so we might not have an issue. If we want to sync it can be an issue since MeshIdentity might not be in another zone, the other option is to put it in status and don't sync status cross zone

```

The resource should also be manually creatable by the user. This enables users to define additional trust domains and allows the local sidecar to accept incoming mTLS traffic from services outside the mesh.
This functionality is essential for:

* Enabling ExternalClients,
* Accepting mTLS connections from external services, and
* Federating multiple separate clusters through shared trust.

**MeshTrust based on MeshIdentity**

> When a user creates a `MeshIdentity`, we will automatically create a corresponding `MeshTrust` resource using the CA and trust domain from that identity. We can implement a dedicated generator that creates the MeshTrust based on the `MeshIdentity`. This option can be disabled and `MeshTrust` might not be created.

Author: `MeshIdentity.provided.trustExtractionDisabled` can disable creation, also we could introduce a flag for controller which will be responsible for this

spec:
  selector:
    dataplane:
      matchLabels: {} # select all
....
---
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: new-identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels:
        app: demo-app
```

The `MeshIdentity` resource’s selector field allows selection based on dataplane labels. Users can define it to apply to either all dataplanes or a specific subset.
If there are multiple (X) identities selecting the same dataplane, the control plane will resolve the conflict by choosing one based on the following order of precedence:

1. Selector specificity – the identity with most specific selector is preferred.
2. Lexicographical order – if selector sizes are equal, the identity with the lexicographically smallest name is chosen.

**Examples**

1. Select only specific dataplane
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
    dataplane:
      matchLabels:
        app: demo-app
```

2. Select all dataplanes
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
    dataplane:
      matchLabels: {}
```

> 3. Select nathing

Reviewer 3: nitpick:

```suggestion
3. Select nothing
```

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
    dataplane: {}
```

or

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
```

or

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
...
```

It seems to be consistant with LabelsMatcher from kubernetes api https://github.com/kubernetes/apimachinery/blob/v0.33.2/pkg/apis/meta/v1/helpers.go#L36-L43.

```golang
func LabelSelectorAsSelector(ps *LabelSelector) (labels.Selector, error) {
	if ps == nil {
		return labels.Nothing(), nil
	}
	if len(ps.MatchLabels)+len(ps.MatchExpressions) == 0 {
		return labels.Everything(), nil
	}
```

> #### How would look the process of switching backends?

Reviewer 3: ```suggestion
#### What would the process of switching backends look like?
```

    trustDomain: "prod.zone-1.kuma.io"
    path: "/ns/{{ .Namespace }}/sa/{{ .ServiceAccount }}"
  provider:
    type: provided
    provided:
      certificate:
        # one of
        secret:
        inline:
        inlineString:
      privateKey:
        # one of
        secret:
        inline:
        inlineString:
      dataplaneCertificate:
        duration: 24h
```
5. Check if works by verifing DataplaneInsights
6. Wait for propagation of values, we caa verify this in `MeshInsights`
7. User remove mTLS section from Mesh object once all dataplanes uses new identity

**Does Envoy supports our certs?**

Yes, if we set `trusted_ca` envoy doesn't do validation of any SAN, hash, etc. https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/ssl#example-configuration

> If only trusted_ca is specified, Envoy will verify the certificate chain of the presented certificate, but not its subject name, hash, etc. Other validation context configuration is typically required depending on the deployment.

SPIFFE validator is experimental but SPIRE uses it. It sets a `custom_validator_config` which allows separation of each bundle to separate trust domain:

```yaml

>                   custom_validator_config:

Reviewer 2: Is new `SPIFFECertValidatorConfig` able to validate our certs that were generated the old way? Is this validator okay with the fact our old certs are not SPIFFE compatible?

>                   custom_validator_config:

Author: Based on the envoy code, we check only first san, and since we put in kuma `spiffe` san as a first record it works correctly

```cc
  for (const GENERAL_NAME* general_name : san_names.get()) {
    if (general_name->type != GEN_URI) {
      continue;
    }

    const std::string san = Utility::generalNameAsString(general_name);
    trust_domain = SPIFFEValidator::extractTrustDomain(san);
    // We can assume that valid SVID has only one URI san.
    break;
  }
```

3. Universal (non-Kubernetes) workflows are SPIFFE-compliant.
4. Support other Identity providers
5. ZoneEgress identity

Nice to have:
1. Kuma control plane can register entries into SPIRE (How can we attest and trust control-plane?)
2. Kuma exposes a federation endpoint.

## Design

### API

#### Default SPIFFE ID

Spire, which is an implementation of SPIFFE, defines a standard way for Subject Alternative Name (SAN) definition. By default, on Kubernetes, it uses the Service Account (SA) and namespace in a path format such as:

```
spiffe://<trustDomain>/ns/<namespace>/sa/<serviceAccount>
```
This default model works for us on Kubernetes but will not be supported in Universal deployments. We therefore need to find a default SPIFFE ID format that works consistently for both Kubernetes and Universal environments.

Unfortunately, we currently lack a native workload identifier in Universal deployments, which prevents us from directly adopting a similar identity model.

**Proposal**

For the current release, we propose defining the default SPIFFE ID for Kubernetes workloads with the following structure:

```
spiffe://<trustDomain>/ns/<namespace>/sa/<serviceAccount>
```

> Trust domain: `{ .Mesh }.{ .Zone }.kuma.io`

Reviewer 3: I'm leaning towards `.mesh.local`

### Migration from mTLS in Mesh to MeshIdentity

1. Mesh has mTLS enabled
```yaml
apiVersion: kuma.io/v1alpha1
kind: Mesh
metadata:
  name: default
spec:
  mtls:
    enabledBackend: ca-1
    backends:
      - name: ca-1
        type: builtin
```
2. User create a new MeshIdentity with example service to verify
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels:
        app: demo-app
  spiffeID: # optional
    trustDomain: "prod.zone-1.mesh.local"
    path: "/ns/{{ .Namespace }}/sa/{{ .ServiceAccount }}"
  bundled: # to extend in KM
    meshTrustCreation: Disabled | Enabled # allows to disable extraction of CA into MeshTrust
    insecureAllowSelfSigned: true
    certificateParameters: # lets make rotation percentage static for the moment
      expiry: 24h
    autogenerate:
      enabled: true
    ca:
      certificate:
        type: File | Secret | EnvVar
        # OneOf
        secret:
        file:
        envVar:
      privateKey:
        type: File | Secret | EnvVar
        # OneOf
        secret:
        file:

>         envVar:

Reviewer 3: do we need this if we have "autogenerate: enabled: true"?

```

In the meantime the `MeshTrust` is created and propagated to each Dataplane.
3. Create MeshTrafficPermissions for accept all traffic
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  name: allow-all
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  from:
  - targetRef:
      kind: Mesh
    default:
      action: Allow
```

> or more specific but it's not a part of this MADR

Reviewer 3: Is it going to be defined in https://github.com/kumahq/kuma/pull/13848 ?

Author: Yes

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTrafficPermission
metadata:
  name: allow-all
  namespace: kuma-demo
  labels:
    kuma.io/mesh: default
spec:
  match:
    rules:
    ....
```
4. Enable MeshIdentity for all service
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels: {}
  spiffeID: # optional
    trustDomain: "prod.zone-1.mesh.local"
    path: "/ns/{{ .Namespace }}/sa/{{ .ServiceAccount }}"

>   provider:

Reviewer 3: isn't it bundled now?

    type: provided
    provided:
      certificate:
        # one of
        secret:
        inline:
        inlineString:
      privateKey:
        # one of
        secret:
        inline:
        inlineString:
      dataplaneCertificate:
        duration: 24h
```
5. Check if works by verifing DataplaneInsights

> 6. Wait for propagation of values, we caa verify this in `MeshInsights`

Reviewer 3: nitpick:

```suggestion
6. Wait for propagation of values, we can verify this in `MeshInsights`
```

7. User remove mTLS section from Mesh object once all dataplanes uses new identity

**Does Envoy supports our certs?**

Yes, if we set `trusted_ca` envoy doesn't do validation of any SAN, hash, etc. https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/ssl#example-configuration

> If only trusted_ca is specified, Envoy will verify the certificate chain of the presented certificate, but not its subject name, hash, etc. Other validation context configuration is typically required depending on the deployment.

> SPIFFE validator is experimental but SPIRE uses it. It sets a `custom_validator_config` which allows separation of each bundle to separate trust domain:

Reviewer 3: What is istio using?

Author: Istio uses `trusted_bundle`

4. Support other Identity providers
5. ZoneEgress identity

Nice to have:
1. Kuma control plane can register entries into SPIRE (How can we attest and trust control-plane?)
2. Kuma exposes a federation endpoint.

## Design

### API

#### Default SPIFFE ID

Spire, which is an implementation of SPIFFE, defines a standard way for Subject Alternative Name (SAN) definition. By default, on Kubernetes, it uses the Service Account (SA) and namespace in a path format such as:

```
spiffe://<trustDomain>/ns/<namespace>/sa/<serviceAccount>
```
This default model works for us on Kubernetes but will not be supported in Universal deployments. We therefore need to find a default SPIFFE ID format that works consistently for both Kubernetes and Universal environments.

Unfortunately, we currently lack a native workload identifier in Universal deployments, which prevents us from directly adopting a similar identity model.

**Proposal**

For the current release, we propose defining the default SPIFFE ID for Kubernetes workloads with the following structure:

```
spiffe://<trustDomain>/ns/<namespace>/sa/<serviceAccount>
```

Trust domain: `{ .Mesh }.{ .Zone }.kuma.io`

> Once a TrustDomain is defined on the Global control plane, we require it to include the zone name in the template. We are going to implement a validator for this case. Also, we need to validate length of each segment where Mesh and Zone has limit of 63 characters.

Author: https://github.com/kumahq/kuma/issues/13054

> Once a TrustDomain is defined on the Global control plane, we require it to include the zone name in the template. We are going to implement a validator for this case. Also, we need to validate length of each segment where Mesh and Zone has limit of 63 characters.

Reviewer 4: that makes sense, can we point to this issue in the MADR so the reader can understand where this limit comes from?

Limitation:

> * Mesh and Zone needs to have max length of 63 characters each

Reviewer 3: it's already a restriction - right? https://github.com/kumahq/kuma/issues/12653

> * Mesh and Zone needs to have max length of 63 characters each

Author: I think covered by this: https://github.com/kumahq/kuma/issues/13054

}
```

**Model**
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels: # move to common structure (golang api)
        app: test 
  spiffeID: # optional
    trustDomain: "default.zone-1.kuma.io" # { .Mesh }.{ .Zone }.kuma.io
    path: "/ns/{{ .Namespace }}/sa/{{ .ServiceAccount }}" # let's use the same templating format as HostnameGenerator
  provider:
    type: Bundled | Spire
    spire:
      agent:
        timeout: 1s
    bundled: # to extend in KM
      meshTrustCreation: Disabled | Enabled # allows to disable extraction of CA into MeshTrust
      insecureAllowSelfSigned: true
      certificateParameters: # lets make rotation percentage static for the moment
        expiry: 24h
      autogenerate:
        enabled: true
      ca:
        certificate:
          type: File | Secret | EnvVar
          # OneOf
          secret:
          file:
          envVar:
        privateKey:
          type: File | Secret | EnvVar
          # OneOf
          secret:
          file:
          envVar:
```

> CP uses the default CA Bundle to validate provided CA. Currently we don't set it, but we can set it additionally by setting `KUMA_GENERAL_TLS_CA_BUNDLE_FILE`.

Reviewer 3: I think I like this option better.

> `insecureAllowSelfSigned` allows preventing users from using self signed CA either provided by the user or autogenrated.

Reviewer 3: nitpick: you've added two dots there Charly - just edited

> `insecureAllowSelfSigned` allows preventing users from using self signed CA either provided by the user or autogenrated.

Reviewer 3: do you mean by default it should be self signed?

> `insecureAllowSelfSigned` allows preventing users from using self signed CA either provided by the user or autogenrated.

Author: Default `insecureAllowSelfSigned : false`

              path: /run/spire/sockets
              type: Directory
        containerPatches: ["spire-agent-socket-patch"]
```

ContainerPatch
```yaml
apiVersion: kuma.io/v1alpha1
kind: ContainerPatch
metadata:
  name: spire-agent-socket-patch
  namespace: kuma-system
spec:
  sidecarPatch:
    - op: add
      path: /volumeMounts/-
      value:
        name: spire-agent-socket
        mountPath: /run/spire/sockets/
        readOnly: true
```

3. We could introduce templating/overlay (out-of-scope for this MADR)

We could introduce overlay to the control-plane injector config.

Example: 

```yaml
runtime:
  kubernetes:
    injector:
        overlay:
            - path: spec.template.spec.volumes.spire-agent-socket
              op: add
              value:
                name: spire-agent-socket
                hostPath:
                  path: /run/spire/sockets
                  type: Directory
            - path: spec.template.spec.containers.kuma-sidecar.volumeMounts.spire-agent-socket
              op: add
              value:
                name: spire-agent-socket
                mountPath: /run/spire/sockets/
                readOnly: true          
```

> #### Define active identity provider

Reviewer: ```suggestion
#### Selecting active identity provider
```

> Previously, defining the active identity for a group of services was done through a single, perhaps less flexible, mechanism. Now, the `MeshIdentity` resource, leveraging its `selector`, offers a more granular and controlled approach to activate specific configurations.

Reviewer: ```suggestion
Previously, selecting the active identity for a group of services was done through a single, perhaps less flexible, mechanism. Now, the `MeshIdentity` resource, leveraging its `selector`, offers a more granular and controlled approach to activate specific configurations.
```

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: old-identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels: {} # select all
....
---
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
metadata:
  name: new-identity
  namespace: kuma-system # only in system namespace
  labels:
    kuma.io/mesh: default
spec:
  selector:
    dataplane:
      matchLabels:
        app: demo-app
```

The `MeshIdentity` resource’s selector field allows selection based on dataplane labels. Users can define it to apply to either all dataplanes or a specific subset.
If there are multiple (X) identities selecting the same dataplane, the control plane will resolve the conflict by choosing one based on the following order of precedence:

> 1. Selector specificity – the identity with most specific selector is preferred.

Reviewer: This should be more rigourous on what specific mean. I think it's by number of keys in the selector.

2. Lexicographical order – if selector sizes are equal, the identity with the lexicographically smallest name is chosen.

**Examples**

1. Select only specific dataplane
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
    dataplane:
      matchLabels:
        app: demo-app
```

2. Select all dataplanes
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
    dataplane:
      matchLabels: {}
```

> 3. Select nothing

Reviewer: I select not more of less specific than select all dp?

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
    dataplane: {}
```

or

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
  selector:
```

or

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshIdentity
...
spec:
...
```

It seems to be consistant with LabelsMatcher from kubernetes api https://github.com/kubernetes/apimachinery/blob/v0.33.2/pkg/apis/meta/v1/helpers.go#L36-L43.

```golang
func LabelSelectorAsSelector(ps *LabelSelector) (labels.Selector, error) {
	if ps == nil {
		return labels.Nothing(), nil
	}
	if len(ps.MatchLabels)+len(ps.MatchExpressions) == 0 {
		return labels.Everything(), nil
	}
```

> #### What would the process of switching backends look like?

Reviewer: what's the value of keeping the same fields? They mention `backends` which are not things that exist anymore in this new world

...
```
Update SpiffeID based on identities
```yaml
...
name: demo-client-65f94cb577-jxrrq-44wfv42db9z44xxf
spec:
  identities:
  - type: SpiffeID
    value: spiffe://mesh-1.zone-1.kuma.io/ns/real-resource-ns/sa/default
  - type: SpiffeID
    value: spiffe://mesh-1.zone-1.kuma.io/ns/real-resource-ns/sa/default/service/demo-client # assume there is different path template
  ports:
  - appProtocol: http
    name: main
    port: 80
    targetPort: main
  selector:
    dataplaneTags:
      app: demo-client
      k8s.kuma.io/namespace: real-resource-ns
  state: Available
status:
  ...
  tls:
    status: Ready
  vips:
  - ip: 10.43.245.159
type: MeshService
```

5. User can remove MeshIdentity `identity-1`
6. MeshTrust should be removed by the user

![Rotation flow](assets/073/rotation-flow.png)

#### Secret naming

We want to use a [KRI](070-resource-identifier.md) for resource naming. Since `MeshIdentity` generates secrets for each dataplane, it should be named based on the KRI to simplify correlation between the Envoy `Secret` and the `MeshIdentity` configuration.

Proposed name:

```yaml
kri_mid_mesh-1_us-east-2_kuma-system_identity-1
```

> Also, we need to name the Spire agent cluster, in this case is it internal or maybe based on MeshIdentity:

Reviewer: ```suggestion
Also, we need to name the Spire agent cluster, in this case it is internal or maybe based on MeshIdentity:
```

Reviewer: I'm confused this say A or B I only see 1 option after...

Proposed name:
```yaml
system_mid_sds-spire-agent
```

> [!WARNING]

> > The naming is different when using SPIRE, since SPIRE uses SPIFFEID or `default` as a certificate name

Reviewer: ```suggestion
> The naming of secrets is different when using SPIRE, since SPIRE uses SPIFFEID or `default` as a certificate name
```

spec:
  selector:
    dataplane:
      matchLabels: # move to common structure (golang api)
        app: test 
  spiffeID: # optional
    trustDomain: "default.zone-1.kuma.io" # { .Mesh }.{ .Zone }.kuma.io
    path: "/ns/{{ .Namespace }}/sa/{{ .ServiceAccount }}" # let's use the same templating format as HostnameGenerator
  provider:
    type: Bundled | Spire
    spire:
      agent:
        timeout: 1s
    bundled: # to extend in KM
      meshTrustCreation: Disabled | Enabled # allows to disable extraction of CA into MeshTrust

>       insecureAllowSelfSigned: true

Reviewer 3: Does it make sense to do an enum? 🤔 

`insecureSelfSigned: allowed | disallowed`
